File: electron/main/ai/openai.ts
import {
  AiProvider,
  CompleteRequest,
  EventSink,
  JsonValue,
  ModelInfo,
  estimateCost,
} from './provider'
import { consumeServerSentEvents } from './sse'

export class OpenAiProvider implements AiProvider {
  private apiKey: string | null
  private baseUrl: string
  private costLabel: string

  constructor(apiKey: string | null, baseUrl: string, costLabel: string) {
    this.apiKey = apiKey
    this.baseUrl = baseUrl.replace(/\/+$/, '')
    this.costLabel = costLabel
  }

  private headers(): Record<string, string> {
    const h: Record<string, string> = { 'content-type': 'application/json' }
    if (this.apiKey && this.apiKey.length > 0) {
      h['authorization'] = `Bearer ${this.apiKey}`
    }
    return h
  }

  async streamComplete(
    req: CompleteRequest,
    sink: EventSink,
    signal: AbortSignal
  ): Promise<void> {
    const messages: Array<{ role: string; content: string }> = []
    if (req.system != null) {
      messages.push({ role: 'system', content: req.system })
    }
    for (const m of req.messages) {
      messages.push({ role: m.role, content: m.content })
    }
    const body: { [k: string]: JsonValue } = {
      model: req.model,
      messages,
      stream: true,
      stream_options: { include_usage: true },
    }
    if (req.temperature != null) body.temperature = req.temperature
    if (req.maxTokens != null) body.max_tokens = req.maxTokens

    let resp: Response
    try {
      resp = await fetch(`${this.baseUrl}/chat/completions`, {
        method: 'POST',
        headers: { ...this.headers(), accept: 'text/event-stream' },
        body: JSON.stringify(body),
        signal,
      })
    } catch (e) {
      if (signal.aborted) return
      throw new Error(`${this.costLabel}-compat request: ${e instanceof Error ? e.message : String(e)}`)
    }

    if (!resp.ok) {
      const text = await resp.text().catch(() => '')
      throw new Error(`${this.costLabel} ${resp.status}: ${text}`)
    }
    if (!resp.body) {
      throw new Error(`${this.costLabel}: empty response body`)
    }

    let inTokens = 0
    let outTokens = 0
    let stopped = false

    await consumeServerSentEvents({
      body: resp.body,
      signal,
      errorPrefix: 'read sse',
      onEvent: (event) => {
        if (event.kind === 'data') {
          if (event.payload === '[DONE]') {
            const cost = estimateCost(this.costLabel, req.model, inTokens, outTokens)
            sink({ kind: 'done', value: { inTokens, outTokens, costUsd: cost } })
            stopped = true
            return true
          }
          return
        }
        const v = event.value
        const choices = v.choices as Array<{ [k: string]: JsonValue }> | undefined
        if (Array.isArray(choices)) {
          for (const choice of choices) {
            const delta = choice.delta as { [k: string]: JsonValue } | undefined
            const text = delta?.content
            if (typeof text === 'string' && text.length > 0) {
              sink({ kind: 'delta', value: text })
            }
          }
        }
        const u = v.usage as { [k: string]: JsonValue } | undefined
        if (u) {
          if (typeof u.prompt_tokens === 'number') inTokens = u.prompt_tokens as number
          if (typeof u.completion_tokens === 'number') outTokens = u.completion_tokens as number
        }
      },
    })

    if (stopped) return
    const cost = estimateCost(this.costLabel, req.model, inTokens, outTokens)
    sink({ kind: 'done', value: { inTokens, outTokens, costUsd: cost } })
  }

  async listModels(): Promise<ModelInfo[]> {
    const resp = await fetch(`${this.baseUrl}/models`, {
      headers: this.headers(),
    })
    if (!resp.ok) {
      const text = await resp.text().catch(() => '')
      throw new Error(`${this.costLabel} models ${resp.status}: ${text}`)
    }
    const parsed = (await resp.json()) as { data: Array<{ id: string }> }
    return parsed.data.map((m) => ({ id: m.id, name: m.id }))
  }
}

File: src-tauri/src/ai/openai.rs
use std::io::{BufRead, BufReader};
use std::sync::atomic::{AtomicBool, Ordering};
use std::time::Duration;

use anyhow::{anyhow, Context, Result};
use reqwest::blocking::{Client, RequestBuilder};
use serde::Deserialize;
use serde_json::json;

use super::provider::{
    estimate_cost, AiEvent, AiProvider, CompleteRequest, EventSink, ModelInfo, Usage,
};

pub struct OpenAiProvider {
    api_key: Option<String>,
    base_url: String,
    cost_label: &'static str,
    client: Client,
}

impl OpenAiProvider {
    pub fn new(api_key: Option<String>, base_url: String, cost_label: &'static str) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(300))
            .build()
            .expect("reqwest client");
        Self {
            api_key,
            base_url: base_url.trim_end_matches('/').to_string(),
            cost_label,
            client,
        }
    }

    fn auth(&self, builder: RequestBuilder) -> RequestBuilder {
        match &self.api_key {
            Some(k) if !k.is_empty() => builder.bearer_auth(k),
            _ => builder,
        }
    }
}

impl AiProvider for OpenAiProvider {
    fn stream_complete(
        &self,
        req: &CompleteRequest,
        sink: EventSink,
        cancel: &AtomicBool,
    ) -> Result<()> {
        let mut messages: Vec<serde_json::Value> = Vec::new();
        if let Some(sys) = &req.system {
            messages.push(json!({"role": "system", "content": sys}));
        }
        for m in &req.messages {
            messages.push(json!({"role": m.role, "content": m.content}));
        }
        let mut body = json!({
            "model": req.model,
            "messages": messages,
            "stream": true,
            "stream_options": { "include_usage": true },
        });
        if let Some(t) = req.temperature {
            body["temperature"] = json!(t);
        }
        if let Some(mt) = req.max_tokens {
            body["max_tokens"] = json!(mt);
        }

        let resp = self
            .auth(
                self.client
                    .post(format!("{}/chat/completions", self.base_url))
                    .header("content-type", "application/json"),
            )
            .json(&body)
            .send()
            .context("openai-compat request")?;

        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().unwrap_or_default();
            return Err(anyhow!("{} {}: {}", self.cost_label, status, text));
        }

        let mut reader = BufReader::new(resp);
        let mut line = String::new();
        let mut in_tokens: u32 = 0;
        let mut out_tokens: u32 = 0;

        loop {
            if cancel.load(Ordering::SeqCst) {
                return Ok(());
            }
            line.clear();
            let n = reader.read_line(&mut line).context("read sse line")?;
            if n == 0 {
                break;
            }
            let trimmed = line.trim_end_matches(['\r', '\n']);
            let payload = match trimmed.strip_prefix("data: ") {
                Some(p) => p,
                None => continue,
            };
            if payload == "[DONE]" {
                break;
            }
            let v: serde_json::Value = match serde_json::from_str(payload) {
                Ok(v) => v,
                Err(_) => continue,
            };
            if let Some(choices) = v.get("choices").and_then(|c| c.as_array()) {
                for choice in choices {
                    if let Some(text) = choice
                        .get("delta")
                        .and_then(|d| d.get("content"))
                        .and_then(|c| c.as_str())
                    {
                        if !text.is_empty() {
                            sink(AiEvent::Delta(text.to_string()));
                        }
                    }
                }
            }
            if let Some(u) = v.get("usage") {
                if let Some(p) = u.get("prompt_tokens").and_then(|x| x.as_u64()) {
                    in_tokens = p as u32;
                }
                if let Some(c) = u.get("completion_tokens").and_then(|x| x.as_u64()) {
                    out_tokens = c as u32;
                }
            }
        }

        let cost = estimate_cost(self.cost_label, &req.model, in_tokens, out_tokens);
        sink(AiEvent::Done(Usage {
            in_tokens,
            out_tokens,
            cost_usd: cost,
        }));
        Ok(())
    }

    fn list_models(&self) -> Result<Vec<ModelInfo>> {
        #[derive(Deserialize)]
        struct ModelsResp {
            data: Vec<ModelEntry>,
        }
        #[derive(Deserialize)]
        struct ModelEntry {
            id: String,
        }
        let resp = self
            .auth(self.client.get(format!("{}/models", self.base_url)))
            .send()
            .context("openai-compat models")?;
        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().unwrap_or_default();
            return Err(anyhow!("{} models {}: {}", self.cost_label, status, text));
        }
        let parsed: ModelsResp = resp.json().context("parse models")?;
        Ok(parsed
            .data
            .into_iter()
            .map(|m| ModelInfo {
                name: m.id.clone(),
                id: m.id,
            })
            .collect())
    }
}
